core/blockchain.go

```diff
@@ -1139,18 +1159,45 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, []
 		//   2. The block is stored as a sidechain, and is lying about it's stateroot, and passes a stateroot
 		// 	    from the canonical chain, which has not been verified.
 		// Skip all known blocks that are behind us
-		current := bc.CurrentBlock().NumberU64()
-		for block != nil && err == ErrKnownBlock && current >= block.NumberU64() {
+		var (
+			current  = bc.CurrentBlock()
+			localTd  = bc.GetTd(current.Hash(), current.NumberU64())
+			externTd = bc.GetTd(block.ParentHash(), block.NumberU64()-1) // The first block can't be nil
+		)
+		externTd = new(big.Int).Add(externTd, block.Difficulty())
+		for block != nil && err == ErrKnownBlock && localTd.Cmp(externTd) >= 0 {
 			stats.ignored++
 			block, err = it.next()
+
+			if block != nil {
+				externTd = new(big.Int).Add(externTd, block.Difficulty())
+			}
+		}
```

Comment:
I think you can simplify this code segment:

```go
for block != nil && err == ErrKnownBlock {
	externTd = new(big.Int).Add(externTd, block.Difficulty())
	if localTd.Cmp(externTd) < 0 {
		break;
	}
	stats.ignored++
	block, err = it.next()
}
```
```

```

Reply:
i.e. you don't need the same externTd += block.Diff calculation both outside and inside the for loop. One seems enough to me. Maybe I'm wrong.